src/io/structure/layers_yaml.rs

```diff
@@ -224,11 +233,17 @@ fn assemble_from_cereal(cereal: self::cereal::Root) -> FailResult<Assemble>
             Lattice::new(&full_lattice),
             CoordsKind::Carts(superstructure.to_carts()),
         );
+
         // FIXME this reduction is just a bandaid for the above-mentioned issue.
         //       (taking unique positions in the diagonal layer supercells and mapping
         //        them into the cell that we generally use for the structure)
         superstructure.reduce_positions();
         fracs_in_plane.push(superstructure.to_fracs());
+
+        for frac in 0..superstructure.num_atoms(){
```

Comment:
In rsp2 I have been careful to make sure that all logic that depends on the order of atoms in a supercell goes through SupercellToken.  The correct way to do this would be to change [this line](https://github.com/ExpHP/rsp2/blob/466e6165be933c7cdef31ccde91197273167891f/src/io/structure/layers_yaml.rs#L229) above it to save the SupercellToken to a variable `sc`:

```rust
let (superstructure, sc) = rsp2_structure::supercell::diagonal(layer.repeat).build(&structure);
```

and then instead of this loop, use the [replicate method](https://github.com/ExpHP/rsp2/blob/012eb1d79542eb371bb01b2e2e6a8a52292af210/src/structure/algo/supercell.rs#L171-L176)

```rust
atoms.extend(sc.replicate(&layer_atoms));
```

(In fact, I believe the current loop would actually be wrong; IIRC rsp2's supercells generate atoms in the order `A A ... A A B B ... B B C C ... C C` rather than `A B C A B C ...`!  This did not affect you since your layers each have only one atom.)